### PR TITLE
ast/If: change if semantics to always return result. unit being the default.

### PR DESCRIPTION
**Pull request by michaellilltokiwa**
I think `unit` is correct here since further down in the code:
```java
private AbstractType typeFromIfOrElse()
  {
    AbstractType result;
    if (hasUntakenElseBranch())
      {
        result = Types.resolved.t_unit;
      }
...
```

### CODE REVIEW

**Comment by fridis** [comment]
The original comment is wrong and your correction matches the behaviour. But reviewing the code of `typeFromIfOrElse`

```
    if (hasUntakenElseBranch())
      {
        result = Types.resolved.t_unit;
      }
    else
      {
        result = Types.resolved.t_void;
        Iterator<Expr> it = branches();
        while (it.hasNext())
          {
            var t = it.next().typeForFeatureResultTypeInferencing();
            result = result == null || t == null ? null : result.union(t);
          }
      }
```
I think this should be
```
    if (hasUntakenElseBranch())
      {
        result = Types.resolved.t_unit;
      }
    else
      {
        result = Types.resolved.t_void;
      }
    Iterator<Expr> it = branches();
    while (it.hasNext())
      {
        var t = it.next().typeForFeatureResultTypeInferencing();
        result = result == null || t == null ? null : result.union(t);
      }
```
Can you try if the tests run fine with this change? Is this related to #536?

**Comment by michaellilltokiwa** [comment]
@fridis I reworked the PR and changed the semantics of `if` to always return a result (default=`unit`) even when there is no explicit else branch.

File: src/dev/flang/ast/If.java [hunk 1]
```diff
@@ -407,20 +375,42 @@ public void propagateExpectedType(Resolution res, AbstractFeature outer)
...
-      elseBlock != null ? elseBlock :
-      elseIf    != null ? elseIf    : new Block(pos(), new List<>());
+    if (PRECONDITIONS) require
+      (elseBlock != null || elseIf != null);
+
+    if (elseBlock != null)
+      {
+        return elseBlock;
+      }
+    return elseIf;
```
Comment (fridis) [inline src/dev/flang/ast/If.java:413]: No a fan of `return` in nested blocks, use 
```
return 
  elseBlock != null ? elseBlock 
                    : elseIf;
```
instead (even if `? :` for many developers needs some time to get used to...).

**Review by fridis — COMMENTED** [review]
Very nice patch, love to see `If.java` getting 25 lines shorter!
Just minor style nit-picking...

**Pull request merged**